### PR TITLE
Save Disposition better

### LINKED ISSUE #12797: Handle Disposition Updating and Saving better
### Background

We made the decision to only send changed fields in update requests so we could better implement virtual hearings logic that only happens if a field was changed. In order to do this we could have either:

1. stored updated values in a separate state from initial states on update, something like
```
<input value={hearing.updated.field || hearing.field} update={(val) => update({ field: val }) />
```
2. Store the entire initial state of the hearing and compare the initial value and changed value at the time of the request

We went with option 2 because it was more easily integrated with the various ways we've implemented forms.

### Issue

On the daily docket, the initial state for the hearing is stored on the `DailyDocketRow`
https://github.com/department-of-veterans-affairs/caseflow/blob/1b6879a99a42922cea5940e089e2f5ea510e0576/client/app/hearings/components/dailyDocket/DailyDocketRow.jsx#L62-L64

and then the comparison is done when the hearing is saved
https://github.com/department-of-veterans-affairs/caseflow/blob/1b6879a99a42922cea5940e089e2f5ea510e0576/client/app/hearings/components/dailyDocket/DailyDocketRow.jsx#L156-L174

This works fine except in the instance when the hearing is `cancelled` or `postponed`. "Currently Scheduled" and "Previously Scheduled" split hearings where the dispositions is null, `held` or `no_show` from those that are `cancelled` or `postponed` into separate `DailyDocketRows`:
https://github.com/department-of-veterans-affairs/caseflow/blob/44405358f7fe40e8d8175ad6e655ed21679690ca/client/app/hearings/components/dailyDocket/DailyDocket.jsx#L205-L212
https://github.com/department-of-veterans-affairs/caseflow/blob/44405358f7fe40e8d8175ad6e655ed21679690ca/client/app/hearings/components/dailyDocket/DailyDocket.jsx#L222-L230

This would be fine except the dropdown that controls disposition opens a modal before save:
https://github.com/department-of-veterans-affairs/caseflow/blob/44405358f7fe40e8d8175ad6e655ed21679690ca/client/app/hearings/components/dailyDocket/DailyDocketRowInputs.jsx#L29-L46

This means that when the currently scheduled hearings are changed to postponed, they are unmounted / removed from the first set of `DailyDocketRows` and mounted on the second, the modal opens, and when the user goes to `saveHearing`, the row no longer exists.

### Solutions
Since this issue only occurs when we're changing the disposition dropdown, we shouldn't complicate the entire save logic for one instance. Let's just change the disposition dropdown to handle this scenario:

```javascript
    onChange={(option) => {
      if (!option) {
        return;
      }

      const fromDisposition = hearing.disposition;
      const toDisposition = option.value;

      openDispositionModal({
        hearing,
        fromDisposition,
        toDisposition,
        onConfirm: saveThenUpdateDisposition
    }}
```
where `saveThenUpdateDisposition` is a slight variation on `saveHearing`, something like
```javascript
// class method on DailyDocketRow.jsx
saveThenUpdateDisposition = (toDisposition) => {
  const hearingWithDisp = { 
    ...this.props.hearing,
    disposition: toDisposition
  };

  const hearing = deepDiff(initialState, hearingWithDisp);

  this.props.saveHearing(hearing.externalId, hearing).then(() => {
    // update redux
    update(hearingWithDisp);
    // only update initialState if hearing isn't postponed or cancelled
    if (['postponed', 'cancelled'].indexOf(toDisp) === -1) {
     this.setState({ initialState: hearingWithDisp });
    }
  });
}
```

### PR DESCRIPTION
**Pull request by aplybeah**
Resolves #12797 

Changes the flow when hearing dispositions get updated on the `Daily Docket` page.
> Before:
* When users clicked `Confirm` on the modal, the disposition would save and the user wouldn't have the chance to make other edits.

> After:
* Disposition will only save once the `Save` button is clicked.

> Nit: 
*  I was unable to find where the `You have successully updated...` alert is triggered. I feel that there are too many alerts now, and that can lead to confusion. (ongoing)

### CODE REVIEW

**Comment by ferristseng** [comment]
This introduces a regression, but I'm not sure why. It looks like the payload that gets sent to the server correctly, but the UI doesn't get updated. In fact, it looks like it gets reverted? It's hard to describe, so I captured it in a GIF:

out

**Comment by ferristseng** [comment]
@tomas-nava Tagged you as a reviewer.

I think this is supposed to fix an issue where when you update the disposition from the dropdown, it updates the hearing disposition in the state, and pushes the daily docket row into the bottom section before you hit save.

**Comment by sahalliburton** [comment]
Also addresses showing save button only when data changed: [#12765 ](https://github.com/department-of-veterans-affairs/caseflow/issues/12765)

File: client/app/hearings/components/dailyDocket/DailyDocketModals.jsx [hunk 1]
```diff
@@ -89,67 +97,79 @@ AodModal.propTypes = {
...
+      <div className="cf-modal-scroll">
+        <Modal
+          closeHandler={onCancel}
+          confirmButton={this.confirmButton()}
+          cancelButton={this.cancelButton()}
+          title={this.COPY[hearingType].title}
+        >
+          <div>
+            <p>
+              Previous Disposition:{' '}
```
Comment (ferristseng) [inline client/app/hearings/components/dailyDocket/DailyDocketModals.jsx:163]: This syntax reads really strangely to me -- Is this something the linter put in?

File: client/app/hearings/utils.js [hunk 2]
```diff
@@ -34,60 +35,110 @@ export const getWorksheetAppealsAndIssues = (worksheet) => {
...
+  // Determine whether the initial value is falsy
+  const falsy = init === null ? false : init;
+
+  // Handle the value comparison
+  switch (current) {
+  // Empty strings should be treated the same as false and null
+  case '':
+  case false:
+    return current != falsy;
+    // Default to compare the initial with the current value
```
Comment (ferristseng) [inline client/app/hearings/utils.js:108]: Should this comment have the same indent as the `default`

File: client/app/hearings/utils.js [hunk 3]
```diff
@@ -34,60 +35,110 @@ export const getWorksheetAppealsAndIssues = (worksheet) => {
...
-export const virtualHearingRoleForUser = (user, hearing) => (
-  user.userCanAssignHearingSchedule || user.userId === hearing.judgeId ?
-    VIRTUAL_HEARING_HOST :
-    VIRTUAL_HEARING_GUEST
-)
+/**
+ * Method to override falsy values for comparison
+ * @param {*} init -- Initial value to compare against
+ * @param {*} current -- Current Value to compare
+ * @returns {boolean} -- Whether the object has been changed
```
Comment (ferristseng) [inline client/app/hearings/utils.js:96]: Nice! Great documentation!

File: client/app/hearings/components/dailyDocket/DailyDocketRow.jsx [hunk 4]
```diff
@@ -70,23 +76,30 @@ class DailyDocketRow extends React.Component {
...
       edited: false,
+      editedFields: [],
       virtualHearingModalActive: false,
       startPolling: null
     };
   }
 
   update = (values) => {
+    // Determine whether the form has been edited and which fields
+    const { edited, editedFields } = handleEdit(this.state.initialState, values, this.state.editedFields);
```
Comment (ferristseng) [inline client/app/hearings/components/dailyDocket/DailyDocketRow.jsx:87]: Cool solution to determining if fields were edited!

File: client/app/hearings/components/dailyDocket/DailyDocketModals.jsx (hunk 1 above)
Comment (sahalliburton) [reply client/app/hearings/components/dailyDocket/DailyDocketModals.jsx:163]: @ferristseng the syntax is a bit strange, it took me to get used to it at first, however this was applied by the linter based on the eslint config and it is React best practice for adding spaces after text in JSX: https://stackoverflow.com/questions/40264084/best-practice-when-adding-whitespace-in-jsx#answer-41142436

File: client/app/hearings/utils.js (hunk 2 above)
Comment (sahalliburton) [reply client/app/hearings/utils.js]: thanks for catching this! not sure why it was different

File: client/app/hearings/components/dailyDocket/DailyDocketModals.jsx (hunk 1 above)
Comment (ferristseng) [reply client/app/hearings/components/dailyDocket/DailyDocketModals.jsx:163]: Cool, thanks for the link!

**Pull request merged**